Core/src/org/sleuthkit/autopsy/centralrepository/datamodel/CorrelationAttributeNormalizer.java

```diff
@@ -0,0 +1,163 @@
+                return CorrelationAttributeNormalizer.normalize(type, data);
+            } else {
+                throw new CorrelationAttributeNormalizationException(String.format("Given attributeTypeId did not correspond to any known Attribute: %s", attributeTypeId));
+            }
+        } catch (EamDbException ex) {
+            throw new CorrelationAttributeNormalizationException(ex);
+        }
+    }
+
+    /**
+     * Verify MD5 is the correct length and values. Make lower case.
+     */
+    private static String normalizeMd5(String data) throws CorrelationAttributeNormalizationException {
+        final String errorMessage = "Data purporting to be an MD5 was found not to comform to expected format: %s";
+        if(data == null){
+            throw new CorrelationAttributeNormalizationException(String.format(errorMessage, data));
+        }
+        final String validMd5Regex = "^[a-f0-9]{32}$";
+        final String dataLowered = data.toLowerCase();
+        if(dataLowered.matches(validMd5Regex)){
+            return dataLowered;
+        } else {
+            throw new CorrelationAttributeNormalizationException(String.format(errorMessage, data));
+        }
+    }
+
+    /**
+     * Verify there are no slashes or invalid domain name characters (such as '?' or \: ). Normalize to lower case.
+     */
+    private static String normalizeDomain(String data) throws CorrelationAttributeNormalizationException {
+        DomainValidator validator = DomainValidator.getInstance(true);
+        if(validator.isValid(data)){
+            return data.toLowerCase();
+        } else {
+            if(data == null){
+                throw new CorrelationAttributeNormalizationException("Data was expected to be a valid domain: null"); 
+            }
+            final String validIpAddressRegex = "^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$";
+            if(data.matches(validIpAddressRegex)){
+                return data;
+            } else {
+                throw new CorrelationAttributeNormalizationException(String.format("Data was expected to be a valid domain: %s", data));
+            }
+        }
+    }
+
+    /**
+     *  Verify that there is an '@' and no invalid characters. Should normalize to lower case.
+     */
+    private static String normalizeEmail(String data) throws CorrelationAttributeNormalizationException {
+        EmailValidator validator = EmailValidator.getInstance(true, true);
+        if(validator.isValid(data)){
+            return data.toLowerCase();
+        } else {
+            throw new CorrelationAttributeNormalizationException(String.format("Data was expected to be a valid email address: %s", data));
+        }
+    }
+
+    /**
+     * Verify it is only numbers and '+'. Strip spaces, dashes, and parentheses.
+     */
+    private static String normalizePhone(String data) throws CorrelationAttributeNormalizationException {
```

Comment:
This and normalizeUsbId should have null checks. Alternately, you could do this up front before the switch statement in normalize().

Reply:
I think the normalize email one actually gracefully handles null in the library I'm using - regardless I just put the check up front before the switch just for simplicity.

Reply:
I was referring to the phone number one - it threw an npe